### PR TITLE
Notes improvements

### PR DESCRIPTION
**Pull request by Timo614**
### References Issues:
References: #1302

### Description:
This branch fixes the associated issue by providing new endpoints for managing notes for an order. There's a get endpoint for fetching all notes of orders, a create endpoint for allowing certain org level users to make notes, and an endpoint for admins, org admins, and org owners to destroy notes in a soft way.

Permissions:
- BoxOffice, can read and write
- DoorPerson, can read and write
- Promoter, can read and write
- PromoterReadOnly, can read
- Org Member, can read and write 
- Org Owner/OrgAdmin/Admin, can read, write, and delete

Endpoints added (scopes in parenthesis):
- DELETE `/notes/{id}` (NoteDelete)
- GET `/orders/{id}/notes` (NoteRead, NoteDelete for filter_deleted=false)
- POST `/orders/{id}/notes` (NoteCreate)

Note:
- Note was removed from DisplayOrder 
- Removed the orders patch endpoint (the logs show it hasn't been used to date) in favor of the new endpoint
- If filter_deleted is passed in and set to false an auth check is done to confirm the user can access the NoteDelete scope -- if they can't it returns unauthorized with normal logging

#### Examples

Fetching notes for an order that doesn't have any yet:
```pg
select * from notes where main_table = 'Orders' and main_id = 'cb5ff12c-fd07-4ed8-97b0-b67ad57f0de9';
 id | note | main_table | main_id | deleted_by | deleted_at | created_by | created_at | updated_at 
----+------+------------+---------+------------+------------+------------+------------+------------
(0 rows)
```

```js
curl 'http://localhost:9000/orders/cb5ff12c-fd07-4ed8-97b0-b67ad57f0de9/notes' -H 'Accept: application/json, text/plain, */*' -H 'Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiI4NDkyZjc5ZC05ZGNiLTQyMjgtYjQ3Zi0wMjhiNGFiY2FjN2YiLCJpc3MiOiJiZy1zYW1wbGUtaXNzdWVyIiwiZXhwIjoxNTYwMzU1NDE4fQ.592ISkxYB3EPyP6NAkPib6v_BURB0h-EZZ9GbqiEWEM' 
{"data":[],"paging":{"page":0,"limit":100,"sort":"","dir":"Asc","total":0,"tags":{"filter_deleted":true}}}
```

Adding new note:
```js
curl 'http://localhost:9000/orders/cb5ff12c-fd07-4ed8-97b0-b67ad57f0de9/notes' -H 'Accept: application/json, text/plain, */*' -H 'Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiI4NDkyZjc5ZC05ZGNiLTQyMjgtYjQ3Zi0wMjhiNGFiY2FjN2YiLCJpc3MiOiJiZy1zYW1wbGUtaXNzdWVyIiwiZXhwIjoxNTYwMzU1NDE4fQ.592ISkxYB3EPyP6NAkPib6v_BURB0h-EZZ9GbqiEWEM' --request POST --data '{"note":"Will be at the box office or some other note"}'  --header "Content-Type: application/json"
{"created_at":"2019-06-12T15:49:05.789629","created_by":"8492f79d-9dcb-4228-b47f-028b4abcac7f","deleted_at":null,"deleted_by":null,"id":"2a83eba4-85e2-46e2-a591-631abd74e515","main_id":"cb5ff12c-fd07-4ed8-97b0-b67ad57f0de9","main_table":"Orders","note":"Will be at the box office or some other note","updated_at":"2019-06-12T15:49:05.789629"}
```

```pg
select * from notes where main_table = 'Orders' and main_id = 'cb5ff12c-fd07-4ed8-97b0-b67ad57f0de9';
                  id                  |                     note                     | main_table |               main_id                | deleted_by | deleted_at |              created_by              |         created_at         |         updated_at         
--------------------------------------+----------------------------------------------+------------+--------------------------------------+------------+------------+--------------------------------------+----------------------------+----------------------------
 2a83eba4-85e2-46e2-a591-631abd74e515 | Will be at the box office or some other note | Orders     | cb5ff12c-fd07-4ed8-97b0-b67ad57f0de9 |            |            | 8492f79d-9dcb-4228-b47f-028b4abcac7f | 2019-06-12 15:49:05.789629 | 2019-06-12 15:49:05.789629
(1 row)
```

```js
curl 'http://localhost:9000/orders/cb5ff12c-fd07-4ed8-97b0-b67ad57f0de9/notes' -H 'Accept: application/json, text/plain, */*' -H 'Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiI4NDkyZjc5ZC05ZGNiLTQyMjgtYjQ3Zi0wMjhiNGFiY2FjN2YiLCJpc3MiOiJiZy1zYW1wbGUtaXNzdWVyIiwiZXhwIjoxNTYwMzU1NDE4fQ.592ISkxYB3EPyP6NAkPib6v_BURB0h-EZZ9GbqiEWEM' 
{"data":[{"id":"2a83eba4-85e2-46e2-a591-631abd74e515","note":"Will be at the box office or some other note","main_table":"Orders","main_id":"cb5ff12c-fd07-4ed8-97b0-b67ad57f0de9","deleted_by":null,"deleted_at":null,"created_by":"8492f79d-9dcb-4228-b47f-028b4abcac7f","created_at":"2019-06-12T15:49:05.789629","updated_at":"2019-06-12T15:49:05.789629"}],"paging":{"page":0,"limit":100,"sort":"","dir":"Asc","total":1,"tags":{"filter_deleted":true}}}
```

Deleting the note now:
```js
curl 'http://localhost:9000/notes/2a83eba4-85e2-46e2-a591-631abd74e515' -H 'Accept: application/json, text/plain, */*' -H 'Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiI4NDkyZjc5ZC05ZGNiLTQyMjgtYjQ3Zi0wMjhiNGFiY2FjN2YiLCJpc3MiOiJiZy1zYW1wbGUtaXNzdWVyIiwiZXhwIjoxNTYwMzU1NDE4fQ.592ISkxYB3EPyP6NAkPib6v_BURB0h-EZZ9GbqiEWEM' --request DELETE
{}
```

```pg
select * from notes where main_table = 'Orders' and main_id = 'cb5ff12c-fd07-4ed8-97b0-b67ad57f0de9';
                  id                  |                     note                     | main_table |               main_id                |              deleted_by              |        deleted_at         |              created_by              |         created_at         |        updated_at         
--------------------------------------+----------------------------------------------+------------+--------------------------------------+--------------------------------------+---------------------------+--------------------------------------+----------------------------+---------------------------
 2a83eba4-85e2-46e2-a591-631abd74e515 | Will be at the box office or some other note | Orders     | cb5ff12c-fd07-4ed8-97b0-b67ad57f0de9 | 8492f79d-9dcb-4228-b47f-028b4abcac7f | 2019-06-12 15:50:22.40012 | 8492f79d-9dcb-4228-b47f-028b4abcac7f | 2019-06-12 15:49:05.789629 | 2019-06-12 15:50:22.40012
(1 row)
```

Trying to fetch records without filter_deleted=false:
```js
curl 'http://localhost:9000/orders/cb5ff12c-fd07-4ed8-97b0-b67ad57f0de9/notes' -H 'Accept: application/json, text/plain, */*' -H 'Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiI4NDkyZjc5ZC05ZGNiLTQyMjgtYjQ3Zi0wMjhiNGFiY2FjN2YiLCJpc3MiOiJiZy1zYW1wbGUtaXNzdWVyIiwiZXhwIjoxNTYwMzU1NDE4fQ.592ISkxYB3EPyP6NAkPib6v_BURB0h-EZZ9GbqiEWEM' 
{"data":[],"paging":{"page":0,"limit":100,"sort":"","dir":"Asc","total":0,"tags":{"filter_deleted":true}}}
```

With filter_deleted=false:
```js
curl 'http://localhost:9000/orders/cb5ff12c-fd07-4ed8-97b0-b67ad57f0de9/notes?filter_deleted=false' -H 'Accept: application/json, text/plain, */*' -H 'Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiI4NDkyZjc5ZC05ZGNiLTQyMjgtYjQ3Zi0wMjhiNGFiY2FjN2YiLCJpc3MiOiJiZy1zYW1wbGUtaXNzdWVyIiwiZXhwIjoxNTYwMzU1NDE4fQ.592ISkxYB3EPyP6NAkPib6v_BURB0h-EZZ9GbqiEWEM' 
{"data":[{"id":"2a83eba4-85e2-46e2-a591-631abd74e515","note":"Will be at the box office or some other note","main_table":"Orders","main_id":"cb5ff12c-fd07-4ed8-97b0-b67ad57f0de9","deleted_by":"8492f79d-9dcb-4228-b47f-028b4abcac7f","deleted_at":"2019-06-12T15:50:22.400120","created_by":"8492f79d-9dcb-4228-b47f-028b4abcac7f","created_at":"2019-06-12T15:49:05.789629","updated_at":"2019-06-12T15:50:22.400120"}],"paging":{"page":0,"limit":100,"sort":"","dir":"Asc","total":1,"tags":{"filter_deleted":false}}}
```

Fetching route with invalid table (404):
```js
curl 'http://localhost:9000/invalid/cb5ff12c-fd07-4ed8-97b0-b67ad57f0de9/notes' -H 'Accept: application/json, text/plain, */*' -H 'Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiI4NDkyZjc5ZC05ZGNiLTQyMjgtYjQ3Zi0wMjhiNGFiY2FjN2YiLCJpc3MiOiJiZy1zYW1wbGUtaXNzdWVyIiwiZXhwIjoxNTYwMzU1NDE4fQ.592ISkxYB3EPyP6NAkPib6v_BURB0h-EZZ9GbqiEWEM' -i
HTTP/1.1 404 Not Found
content-length: 21
content-type: application/json
access-control-allow-origin: http://localhost:3000
access-control-expose-headers: x-app-version
vary: Origin
x-app-version: 1.0.145
date: Wed, 12 Jun 2019 15:51:28 GMT

{"error":"Not found"}
```

Fetching route with valid table but not yet supported from auth perspective:
```js
curl 'http://localhost:9000/users/cb5ff12c-fd07-4ed8-97b0-b67ad57f0de9/notes' -H 'Accept: application/json, text/plain, */*' -H 'Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiI4NDkyZjc5ZC05ZGNiLTQyMjgtYjQ3Zi0wMjhiNGFiY2FjN2YiLCJpc3MiOiJiZy1zYW1wbGUtaXNzdWVyIiwiZXhwIjoxNTYwMzU1NDE4fQ.592ISkxYB3EPyP6NAkPib6v_BURB0h-EZZ9GbqiEWEM' -i
HTTP/1.1 401 Unauthorized
content-length: 55
content-type: application/json
access-control-allow-origin: http://localhost:3000
access-control-expose-headers: x-app-version
vary: Origin
x-app-version: 1.0.145
date: Wed, 12 Jun 2019 15:51:49 GMT

{"error":"User does not have the required permissions"}
```

### Migrations
 * 20190611150434_notes_improvements - creates new table for notes, adds indexes, and converts the old note data from domain events over into new notes

### Environment Variables
 * No change

### CODE REVIEW

File: db/src/models/notes.rs [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+impl NewNote {
+    pub fn commit(&self, conn: &PgConnection) -> Result<Note, DatabaseError> {
+        self.validate()?;
+
+        let note: Note = diesel::insert_into(notes::table)
+            .values(self)
+            .get_result(conn)
+            .to_db_error(ErrorCode::InsertError, "Could insert new note")?;
+
+        for (table, id) in vec![(Tables::Orders, note.order_id), (Tables::Notes, note.id)] {
```
Comment (Krakaw) [inline db/src/models/notes.rs]: Are we sure we want to duplicate the domain event, might land up being a fair amount of data?

File: db/src/models/notes.rs [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        }
+
+        let (notes, record_count): (Vec<Note>, i64) = query
+            .order_by(notes::created_at.desc())
+            .paginate(page as i64)
+            .per_page(limit as i64)
+            .load_and_count_pages(conn)
+            .to_db_error(ErrorCode::QueryError, "Error loading notes")?;
+
+        let mut payload = Payload::from_data(notes, page, limit);
```
Comment (Krakaw) [inline db/src/models/notes.rs]: This is nice and clean!

File: db/src/models/scopes.rs [hunk 3]
```diff
@@ -210,6 +219,7 @@ fn get_scopes_for_role(role: Roles) -> Vec<Scopes> {
                 Scopes::RedeemTicket,
                 Scopes::OrgReadEvents,
                 Scopes::HoldRead,
+                Scopes::NoteRead,
```
Comment (Krakaw) [inline db/src/models/scopes.rs]: @zmelamed would we want the doorperson / boxoffice person to make notes about a customer, such as they were unruly in the line or they should get an upgrade next time etc etc.

Comment (zmelamed) [reply db/src/models/scopes.rs]: Yeah we should allow for that. Great thought thanks.

File: api/src/controllers/notes.rs [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+#[derive(Deserialize, Serialize)]
+pub struct NoteFilterParameters {
+    pub filter_deleted: Option<bool>,
+}
+
+pub fn create(
+    (conn, path, json, auth_user): (Connection, Path<PathParameters>, Json<NewNoteRequest>, User),
+) -> Result<HttpResponse, BigNeonError> {
+    let connection = conn.get();
+    let order = Order::find(path.id, connection)?;
```
Comment (mikethetike) [inline api/src/controllers/notes.rs]: I was thinking of having the notes link to tables dynamically like DomainEvents and DomainActions do. Then we can add notes to any enitity

File: db/src/models/notes.rs (hunk 1 above)
Comment (mikethetike) [reply db/src/models/notes.rs]: I also think one should be sufficient. The other key should be stored in the event_data

**Comment by mikethetike** [comment]
Happy for it to go in as is, but I think the dynamic link to any entity might be very useful

**Comment by Timo614** [comment]
Will adjust this today for the above.

**Comment by Timo614** [comment]
I've updated this branch re: the feedback. The API still works in the same way I just have it scoped for creation to the main_table so as we add support for other tables (and their auth checks) it will just work. Replaced the examples above in the issue text as well for the new format with main_table and main_id.

**Pull request merged**